src/tls/test/main.cpp

```diff
@@ -22,67 +22,205 @@ static const string contents_ =
-  tls::PublicKey pubk(public_key);
-  REQUIRE(pubk.verify(
-    contents.data(), contents.size(), signature.data(), signature.size()));
+  for (const auto curve : supported_curves)
+  {
+    INFO("With curve: " << labels[static_cast<size_t>(curve) - 1]);
+    auto kp = tls::make_key_pair(curve);
+    vector<uint8_t> contents(contents_.begin(), contents_.end());
+    const vector<uint8_t> signature = kp->sign(contents);
+
+    vector<uint8_t> public_key = kp->public_key();
+    auto pubk = tls::make_public_key(public_key);
+    CHECK(pubk->verify(contents, signature));
+  }
 }
 
 TEST_CASE("Sign, fail to verify with bad signature")
 {
-  tls::KeyPair kp;
-  vector<uint8_t> contents(contents_.begin(), contents_.end());
-  vector<uint8_t> signature = kp.sign(contents);
-
-  vector<uint8_t> public_key = kp.public_key();
-  tls::PublicKey pubk(public_key);
-  corrupt(signature);
-  REQUIRE_FALSE(pubk.verify(
-    contents.data(), contents.size(), signature.data(), signature.size()));
+  for (const auto curve : supported_curves)
+  {
+    INFO("With curve: " << labels[static_cast<size_t>(curve) - 1]);
+    auto kp = tls::make_key_pair(curve);
+    vector<uint8_t> contents(contents_.begin(), contents_.end());
+    vector<uint8_t> signature = kp->sign(contents);
+
+    vector<uint8_t> public_key = kp->public_key();
+    auto pubk = tls::make_public_key(public_key);
+    corrupt(signature);
+    CHECK_FALSE(pubk->verify(contents, signature));
+  }
 }
 
 TEST_CASE("Sign, fail to verify with bad contents")
 {
-  tls::KeyPair kp;
-  vector<uint8_t> contents(contents_.begin(), contents_.end());
-  vector<uint8_t> signature = kp.sign(contents);
-
-  vector<uint8_t> public_key = kp.public_key();
-  tls::PublicKey pubk(public_key);
-  corrupt(contents);
-  REQUIRE_FALSE(pubk.verify(
-    contents.data(), contents.size(), signature.data(), signature.size()));
+  for (const auto curve : supported_curves)
+  {
+    INFO("With curve: " << labels[static_cast<size_t>(curve) - 1]);
+    auto kp = tls::make_key_pair(curve);
+    vector<uint8_t> contents(contents_.begin(), contents_.end());
+    vector<uint8_t> signature = kp->sign(contents);
+
+    vector<uint8_t> public_key = kp->public_key();
+    auto pubk = tls::make_public_key(public_key);
+    corrupt(contents);
+    CHECK_FALSE(pubk->verify(contents, signature));
+  }
+}
+
+TEST_CASE("Sign, fail to verify with wrong key on correct curve")
+{
+  for (const auto curve : supported_curves)
+  {
+    INFO("With curve: " << labels[static_cast<size_t>(curve) - 1]);
+    auto kp = tls::make_key_pair(curve);
+    vector<uint8_t> contents(contents_.begin(), contents_.end());
+    vector<uint8_t> signature = kp->sign(contents);
+
+    auto kp2 = tls::make_key_pair(curve);
+    vector<uint8_t> public_key = kp2->public_key();
+    auto pubk = tls::make_public_key(public_key);
+    CHECK_FALSE(pubk->verify(contents, signature));
+  }
+}
+
+TEST_CASE("Sign, fail to verify with wrong key on wrong curve")
```

Comment:
Should we add a test for "wrong key on right curve"?

Reply:
That's the test case above this one

`TEST_CASE("Sign, fail to verify with wrong key on correct curve")`